### DISCUSSION TITLE
Circular progress bar

### DISCUSSION THREAD

**Discussion by wolfgan** [fyne-io/fyne#4038] [Q&A]
I would like to have a circular progress bar. It is not possible to do this on the basis of a circle. I would like to have Pie in the arsenal. Now I'm trying to make progress based on many circles. Approximately in this format:

![image](https://github.com/fyne-io/fyne/assets/7305231/b6b3e0e1-293d-42b1-9037-c05edfe492f8)

**Comment by matwachich** [comment]
Check fyne-x, there is a GIF widget you can use for this.

**Reply by wolfgan** [reply]
An animated GIF is suitable for an endless operation. But what about showing the percentage of completion? Now I have two ideas: use a set of png slides or try to implement it with many circles (as in the screenshot above).
![image](https://github.com/fyne-io/fyne/assets/7305231/2ddf5713-c6a9-416a-aa1d-055a2878755a)

**Comment by wolfgan** [comment]
I was able to do what I want.
https://github.com/fyne-io/fyne/assets/7305231/8e9b2866-cc6b-42f6-a8f2-6e279de0eac0

**Reply by PaulWaldo** [reply]
@Wolfgan is extending the existing `ProgressBar` widget not viable (https://developer.fyne.io/extend/extending-widgets)? I'm not trying to be snarky; I'm genuinely interested.  I'm relatively new to fyne, but I would think overriding `CreateRenderer` and such would allow you to use the existing API, yet change the presentation.

**Reply by Jacalz** [reply]
Well done. That looks awesome. Any plans to contribute that onto [fyne-x](https://github.com/fyne-io/fyne-x)? It can probably be useful for more people :)

**Reply by wolfgan** [reply]
@PaulWaldo  Unfortunately, it doesn't work in this case. You can add new functionality to a widget this way, but not change how it works. Here it was required to completely create a new widget. Actually it was not very difficult, but there were some difficulties. You can't just take and make a new widget in your project, because. some necessary things are inside the package (internal). The only way is to fork.

**Reply by wolfgan** [reply]
@Jacalz  I think yes, I will try to do it when I have time. If you need it right now, then you can see the code of the waters here: 
https://github.com/wolfgan/fyne/blob/master/widget/dots_progressbar.go
https://github.com/wolfgan/fyne/blob/master/widget/dots_progressbar_infinity.go
However, remember, it is not yet sufficiently tested and an error may occur.

**Reply by andydotxyz** [reply]
> @PaulWaldo Unfortunately, it doesn't work in this case. You can add new functionality to a widget this way, but not change how it works.

In this case is it not just changing how it looks - not how it works? In which case you could extend the progress bar and replace the `CreateRenderer` call I think...

**Reply by wolfgan** [reply]
@andydotxyz  This was one of the methods I tried. However, I did not find a way to replace the render that is created automatically and entered into an internal cache that is not accessible. You can see how this is done on the example of ProgressBarInfinity. Now I have an idea how to do it differently, but I suspect that the renderer will be implicitly created anyway. I think I need to look at the source code again.

**Reply by andydotxyz** [reply]
Perhaps you missed the `ExtendBaseWidget` before `Refresh` was called for the first time?

**Reply by wolfgan** [reply]
I call `ExtendBaseWidget`  in methods `CreateRenderer` and `MinSize`.  Perhaps this is not enough. You can [see](https://github.com/wolfgan/fyne/blob/master/widget/dots_progressbar_infinity.go#L206C1-L206C1) how it's done here.

**Reply by PaulWaldo** [reply]
@wolfgan I believe the only place you need to call `ExtendBaseWidget` is in the constructor of your widget.  Once this is done, your widget will be the receiver of the standard widget calls.

**Reply by Jacalz** [reply]
I'd say that @wolfgan did the right thing. We place it there (and usually in the constructor as well) so that people can use the widget created with the struct directly if they so prefer.

**Reply by wolfgan** [reply]
The problem is not in `ExtendBaseWidget`.  The problem is that I cannot access some methods outside the package.

`// Stop the infinite progress bar animation
func (p *DotsProgressBarInfinity) Stop() {
	cache.Renderer(p).(*dotsProgressInfinityRenderer).stop()
}

// Running returns the current state of the infinite progress animation
func (p *DotsProgressBarInfinity) Running() bool {
	return cache.Renderer(p).(*dotsProgressInfinityRenderer).isRunning()
}`

Animation is in `Render.` How to access control without using `cache.Renderer`?
Only because of this I was forced to do Fork. Otherwise, I would make a widget for a fine-x.

**Reply by Jacalz** [reply]
Ah, I see. There might be a case for rewriting that widget to not interact with the renderer. I've never personally been a fan of doing it the way it is currently set up. Is there any specific reason for doing so? @andydotxyz

**Reply by andydotxyz** [reply]
You can extend it, replace those public methods in your own code but call the public api from your own code can't you? (Like a delegate pattern).

Perhaps they should look for the cache of `p.super()` rather than `p`.

Anyhow it should not require a fork of Fyne - your progress bar could replicate the API instead of extending the existing widgets I would think...

**Reply by wolfgan** [reply]
In the end, I got rid of this problem. Rewrite without using animation, replacing it with a running line. You can see [here](https://github.com/wolfgan/fxw)
I save the render on creation. Then, using a cast, you can access its methods. Thus, it was possible not to remove the animation, but I could no longer be stopped :)

**Comment by wolfgan** [comment]
Now it looks like this:

https://github.com/fyne-io/fyne/assets/7305231/a6e8ad40-3108-4a68-bea7-30c1696a166f

**Reply by Jacalz** [reply]
Looks great. Well done
